atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java

```diff
@@ -101,11 +100,43 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
             loadSingleBlockToOutputStream(transaction, id, 0, ios);
             return ios.getInputStream();
         } else {
-            File file = loadToNewTempFile(transaction, id, metadata);
-            return new DeleteOnCloseFileInputStream(file);
+            return makeStream(id, metadata);
         }
     }
 
+    private InputStream makeStream(ID id, StreamMetadata metadata) {
+        int numBlocks = (int) getNumberOfBlocksFromMetadata(metadata);
+        int blocksInMemory = numberOfBlocksThatFitInMemory();
+
+        BlockGetter pageRefresher = new BlockGetter() {
+            @Override
+            public void get(int firstBlock, int numBlocks, OutputStream destination) {
+                txnMgr.runTaskReadOnly(txn -> {
+                    for (int i = 0; i < numBlocks; i++) {
+                        loadSingleBlockToOutputStream(txn, id, firstBlock + i, destination);
+                    }
+                    return null;
+                });
+            }
+
+            @Override
+            public int expectedLength() {
+                return BLOCK_SIZE_IN_BYTES * blocksInMemory;
+            }
+        };
+
+        try {
+            return BlockConsumingInputStream.create(pageRefresher, numBlocks, blocksInMemory);
+        } catch (IOException e) {
+            throw Throwables.throwUncheckedException(e);
+        }
+    }
+
+    private int numberOfBlocksThatFitInMemory() {
+        long blocksInMemory = getInMemoryThreshold() / BLOCK_SIZE_IN_BYTES;
+        return Math.max(1, (int) blocksInMemory);
```

Comment:
On closer inspection, I note that `buffer.length` is an int, so the true maximum number of blocks is `Integer.MAX_VALUE / BLOCK_SIZE_IN_BYTES = 2147`. It's worth adding a test for that, as it's only about 2GB.